Enhance torch.vmap support from inside torch.compile

guilhermeleobas opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116050

This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

guilhermeleobas committed:
Enhance torch.vmap support from inside torch.compile

This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

[ghstack-poisoned]

Skylion007 commented on torch/utils/_pytree.py:
```diff
@@ -752,7 +752,7 @@ def tree_map_(func: Callable[..., Any], tree: PyTree, *rests: PyTree) -> PyTree:
     leaves, treespec = tree_flatten(tree)
     flat_args = [leaves] + [treespec.flatten_up_to(r) for r in rests]
-    deque(map(func, *flat_args), maxlen=0)  # consume and exhaust the iterable
+    tuple(map(func, *flat_args))  # consume and exhaust the iterable
```
This isn't exactly equivalent and will be slower: https://stackoverflow.com/questions/36763084/how-to-efficiently-exhaust-an-iterator-in-a-oneliner

zou3519 replied:
It's OK, we don't care too much about the performance of `tree_map_` right now

zou3519 commented on torch/_dynamo/trace_rules.py:
```diff
@@ -93,6 +94,29 @@
     "torch._nested_tensor_from_mask": SkipFilesVariable,
     "torch._nested_from_padded": SkipFilesVariable,
+    # functorch
+    "torch._functorch.vmap._check_int_or_none": UserFunctionVariable,
+    "torch._functorch.vmap._check_out_dims_is_int_or_int_pytree": UserFunctionVariable,
+    "torch._functorch.vmap._check_randomness_arg": UserFunctionVariable,
+    "torch._functorch.vmap._chunked_vmap": UserFunctionVariable,
+    "torch._functorch.vmap._concat_chunked_outputs": UserFunctionVariable,
+    "torch._functorch.vmap._create_batched_inputs": UserFunctionVariable,
+    "torch._functorch.vmap._flat_vmap": UserFunctionVariable,
+    "torch._functorch.vmap._flatten_chunks_output": UserFunctionVariable,
+    "torch._functorch.vmap._get_chunked_inputs": UserFunctionVariable,
+    "torch._functorch.vmap._get_name": UserFunctionVariable,
+    "torch._functorch.vmap._maybe_remove_batch_dim": UserFunctionVariable,
+    "torch._functorch.vmap._num_outputs": UserFunctionVariable,
+    "torch._functorch.vmap._process_batched_inputs": UserFunctionVariable,
+    "torch._functorch.vmap._unwrap_batched": UserFunctionVariable,
+    "torch._functorch.vmap._validate_and_get_batch_size": UserFunctionVariable,
+    "torch._functorch.vmap.doesnt_support_saved_tensors_hooks": UserFunctionVariable,
+    "torch._functorch.vmap.get_chunk_sizes": UserFunctionVariable,
+    # "torch._functorch.vmap.lazy_load_decompositions": UserFunctionVariable,
+    "torch._functorch.vmap.restore_vmap": UserFunctionVariable,
+    "torch._functorch.vmap.unwrap_batched": UserFunctionVariable,
+    "torch._functorch.vmap.vmap_impl": UserFunctionVariable,
+    "torch._functorch.vmap.wrap_batched": UserFunctionVariable,
```
@yanboliang is there a succinct way to express that we want everything in "torch._functorch.vmap" to be inlined?

yanboliang replied:
Yes, two steps:
* Add ```torch._functorch.vmap``` to https://github.com/pytorch/pytorch/blob/626b7dc847bff19c70b200fa2228bf2b8f4df753/torch/_dynamo/skipfiles.py#L197
* Remove majority ```torch._functorch.vmap``` functions from this list.
  - You have to keep the function that dynamo have implementation in this list, e.g., ```torch._functorch.vmap.vmap_impl```.

zou3519 replied:
I think we already did that, but for some reason it didn't work?

yanboliang replied:
Oh, then you have to do the following two steps:
* Remove these functions from the trace_rules.py list, because it takes precedence over the ```skipfiles.MOD_INLINELIST``` check.
* Wait my next PR to remove ```is_allowed```, which I plan to publish in 1-2 days.

Then these functions would be inlined.

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/skipfiles.py:
```diff
@@ -201,6 +200,7 @@ def _module_dir(m: types.ModuleType):
     "torch._dynamo._trace_wrapped_higher_order_op",
     "torch._dynamo.comptime",
     "torch._dynamo.polyfill",
+    "torch._functorch.vmap",
```
@yanboliang it's already here

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3289,11 +3336,7 @@ def fn(x, y):
-            dict(counters["graph_break"]),
-            {"NYI - torch.func.vmap: kwargs arguments are currently unsupported.": 2},
-        )
+        self.assertEqual(len(counters["graph_break"]), 0)
```
@guilhermeleobas: We should add some explicit tests that check torch.compile(backend="aot_eager") where the function being vmapped over has graph breaks in it. I am not sure this situation works (in theory, it probably doesn't).

Here's an example:
```py
def g(x):
   y = x.sum(0)
   print("a")  # cause a graph break
   z = y.sum(0)

@torch.compile(backend="aot_eager")
def f(x):
    return vmap(g)(x)

x = torch.randn(2, 3, 4, 5)
f(x)
```
I would assume this does not work because AOTAutograd doesn't know what to do when Dynamo passes it a graph that has a BatchedTensor input or output -- it only knows what to do in the situation where Dynamo passes it a graph whose input and outputs are regular Tensors.

In this situation, for now, we would want the entire of vmap(g)(x) to be run in eager-mode PyTorch. Open question on how to actually get there.

zou3519 replied:
Bounced some ideas with Voz. The TL;DR:
1. If there is a graph break, what we want to do is apply torch._dynamo.disable to vmap(g), and then carry on with the graph break. This will prevent Dynamo from looking into that specific vmap(g) again (vmap(g) returns a new function every time it is called)
2. A user might do something like the following, where they save the result of vmap(f):
```
vf = vmap(f)
vf(x)
vf(y)
```
However, this ends up with the problem that if vf(x) graph breaks and falls back to eager-mode PyTorch, we do not necessarily want vf(y) to graph break. One way to resolve this is to have vf(x) actually generate fresh new function `k` every time it is called and invoke k(x), and then the thing we actually torch._dynamo.disable is `k`.

TBH we can do (2) as a follow-up, but this PR should have (1) at the least

guilhermeleobas replied:
I think I got it! There's a `torch._C._dynamo.eval_frame.skip_code` function that one can mark a frame to be skipped. This avoids Dynamo from tracing the frame again in case of a graph break. Will add a couple of tests in a bit.

zou3519 commented on torch/_dynamo/utils.py:
```diff
@@ -1451,6 +1451,9 @@ def get_fake_values_from_nodes(tx, nodes):
         return n.meta["example_value"]
 
+    # unwrap proxy nodes
+    nodes = tree_map_only(fx.Proxy, lambda x: x.node, nodes)
```
Why do we need this?

guilhermeleobas replied:
We don't. I'll remove it.

zou3519 commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -335,7 +336,7 @@ def inner(*flat_args):
             curr_storage = (
                 None
-                if not isinstance(o, torch.Tensor)
+                if not isinstance(o, torch.Tensor) or is_batchedtensor(o)
```
TODO: I'm not sure what's going on in this file

guilhermeleobas replied:
We don't actually need those changes. I'll remove them.

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -102,6 +102,13 @@ def closure_vars(self, tx):
 class UserFunctionVariable(BaseUserFunctionVariable):
     """Some unsupported user-defined global function"""
 
+    @classmethod
+    def create_with_source(cls, value, source):
+        return cls(
+            value,
+            source=source,
+        )
```
Could you explain the changes to UserFunctionVariable?

zou3519 reviewed (commented):
I mostly just have questions, but this looks pretty good to me. Also had a suggestion for a test case (that may expose more problems)

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2916,23 +2920,27 @@ def fn(x):
+        _vmap_increment_nesting = torch._C._functorch._vmap_increment_nesting(3, 'error')
+
+        _add_batch_dim = torch._C._functorch._add_batch_dim(arg, 0, 1);  arg = None
+
```
Another idea for test cases: we should remove all of the skipIfTorchDynamo/xfailIfTorchDynamo from https://github.com/pytorch/pytorch/blob/main/test/functorch/test_vmap.py and fix the failures. It doesn't have to happen in this PR (as long as this PR doesn't add more of them).

guilhermeleobas replied:
Working on this one at the moment.

guilhermeleobas replied:
This is blocked until https://github.com/pytorch/pytorch/issues/116264 gets fixed

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2964,22 +3070,24 @@ def fn(x):
-        vmap_proxy = torch.func.vmap(vmap_body_0, (0,), 0, 'error');  vmap_body_0 = None
-        call = vmap_proxy.__call__(child);  vmap_proxy = child = None
-        return (call,)
+        _saved_tensors_hooks_disable = torch._C._autograd._saved_tensors_hooks_disable("torch.func transforms don't yet support saved tensor hooks. Please open an issue with your use case.")
```
To check: like vmap_increment_nesting, _saved_tensors_hooks_disable is also a context-manager-like thing, did we implement it as a ContextManager? (i.e. does Dynamo undo the state on a graph break inside vmap(f)(x))?

guilhermeleobas replied:
@kshitij12345 implemented support for it in https://github.com/pytorch/pytorch/pull/104869

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2964,22 +3070,24 @@ def fn(x):
-            sum_2 = select.sum(1);  select = None
-            add = sum_1 + sum_2;  sum_1 = sum_2 = None
-            return add
+        _add_batch_dim = torch._C._functorch._add_batch_dim(arg, 0, 1);  arg = None
```
I'm not sure how to check this, but I want to make sure the guards are correct. It looks like we're baking in the level (_vmap_increment_nesting from the previous line) as equal to 0. I would expect that we add a guard for `torch._C._functorch.current_level() is None` or something -- this Dynamo-produced FX graph is only valid if the current functorch level before calling vmap(f) is None.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3266,23 +3392,24 @@ def fn(x):
-        getitem_1 = call[1];  call = None
-        return (getitem, getitem_1)
+        _saved_tensors_hooks_disable = torch._C._autograd._saved_tensors_hooks_disable("torch.func transforms don't yet support saved tensor hooks. Please open an issue with your use case.")
+        _vmap_increment_nesting = torch._C._functorch._vmap_increment_nesting(2, 'error')
```
Why is there 1 _vmap_increment_nesting in this graph (but two _vmap_decrement_nesting)?

zou3519 replied:
Oops i misread.

zou3519 commented on test/functorch/test_vmap.py:
```diff
@@ -1178,6 +1178,7 @@ def f(x, y):
         self.assertEqual(out, f(None, y))
         self.assertEqual(out_dims, (None, None, None))
 
+    @xfailIfTorchDynamo
```
What's going on here?

guilhermeleobas replied:
This does not raise an error under torch.compile

zou3519 replied:
Let's ignore this for now, file an issue for it so we come back to it.

zou3519 commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -148,6 +148,38 @@ def exit(self, tx, *args):
         return x
 
 
+class VmapCtxManagerVariable(ContextWrappingVariable):
```
nit: name it VmapIncrementNestingCtxManagerVariable

zou3519 commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -148,6 +148,38 @@ def exit(self, tx, *args):
+            (batch_size, randomness),
+            {},
+        )
+        return variables.ConstantVariable.create(vmap_level)
```
hmm, this might need a guard

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -323,6 +322,12 @@ def call_function(
+            assert len(args) == 2
+            return VmapCtxManagerVariable.create(
+                tx,
+                [guard_if_dyn(x) for x in args],
```
What's this?

guilhermeleobas replied:
`guard_if_dyn` gets the underlying python constant variable, which in this case are the `batch_size` and `randomness` values.

zou3519 commented on torch/_functorch/vmap.py:
```diff
@@ -390,15 +384,22 @@ def _check_randomness_arg(randomness):
 
 
+@contextlib.contextmanager
+def vmap_nesting(batch_size, randomness):
```
nit: vmap_nesting -> vmap_increment_nesting ?

zou3519 reviewed (commented):
Looks really good, especially if tests pass. Left some comments

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
@zou3519, could you review this PR again? CI is green.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2370,7 +2372,68 @@ def fn(pred, pytree_in):
                 torch.compile(fn, backend="eager")(pred, pytree_in)
 
 
+class HigherOrderOpLoggingTests(LoggingTestCase):
```
Nit: are these really Logging tests?

guilhermeleobas replied:
No! I've changed the class name to `HigherOrderOpVmapGuardTests`

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2370,7 +2372,68 @@ def fn(pred, pytree_in):
+
+
 class FuncTorchHigherOrderOpTests(torch._dynamo.test_case.TestCase):
+    def tearDown(self):
+        # Ensure that in the case of a test failure, the next test won't fail
+        # because of a previous call to _vmap_increment_nesting that wasn't undone
+        # i.e. test_vmap_free_tensor fails when PYTORCH_TEST_WITH_DYNAMO=1
+        # and the call to increment nesting is not undone
+        ci = torch._C._functorch.peek_interpreter_stack()
+        if ci and ci.key() == torch._C._functorch.TransformType.Vmap:
+            torch._C._functorch._vmap_decrement_nesting()
```
This ends up hiding problems, right? Instead, in tearDown, we should assert that the functorch stack is empty.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2946,6 +3009,112 @@ def wrapper_fn(x, y):
+        def g(x):
+            return x.sin()
+
+        @torch.compile(backend="aot_eager")
```
fullgraph=True ?

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3171,32 +3350,37 @@ def fn(x, y):
+        _saved_tensors_hooks_disable_1 = torch._C._autograd._saved_tensors_hooks_disable("torch.func transforms don't yet support saved tensor hooks. Please open an issue with your use case.")
+        _vmap_increment_nesting_1 = torch._C._functorch._vmap_increment_nesting(3, 'error')
+
+        _add_batch_dim_2 = torch._C._functorch._add_batch_dim(arg_8, 1, 2);  arg_8 = None
+        _add_batch_dim_3 = torch._C._functorch._add_batch_dim(arg_9, 1, 2);  arg_9 = None
```
nit: what are the chances we can pass level and bdim as kwargs? No worries if not, it'll just make it easier to read.

guilhermeleobas replied:
It is ok if I address this in a follow-up PR?

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3426,14 +3607,7 @@ def wrapper_fn(x, y):
-                r".*HigherOrderOperator: Mutating a variable not in the current scope \(SideEffects\)": 2
-            },
-        )
+        self.assertEqual(len(counters["graph_break"]), 0)
```
Can we check that some_list == [1, 1]? 

Also, can we add a test for what happens if the function looks like:

```
def f(x, y):
   some_list.append(x)
   return x + y
```

guilhermeleobas replied:
This one doesn't work

```
Traceback (most recent call last):
  File "/home/guilhermeleobas/micromamba/envs/pytorch/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "test/dynamo/test_higher_order_ops.py", line 3631, in test_vmap_side_effects_append_input
    expected = torch.compile(wrapper_fn, backend="aot_eager", fullgraph=False)(x, y)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/eval_frame.py", line 417, in _fn
    return fn(*args, **kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/eval_frame.py", line 580, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/convert_frame.py", line 741, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/convert_frame.py", line 384, in _convert_frame_assert
    return _compile(
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/convert_frame.py", line 643, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/convert_frame.py", line 524, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
    transformations(instructions, code_options)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/convert_frame.py", line 151, in _fn
    return fn(*args, **kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/convert_frame.py", line 489, in transform
    tracer.run()
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/symbolic_convert.py", line 2111, in run
    super().run()
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/symbolic_convert.py", line 780, in run
    and self.step()
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/symbolic_convert.py", line 743, in step
    getattr(self, inst.opname)(inst)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/symbolic_convert.py", line 2229, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/output_graph.py", line 940, in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
  File "/home/guilhermeleobas/micromamba/envs/pytorch/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/output_graph.py", line 1085, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/output_graph.py", line 1157, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/output_graph.py", line 1138, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/guilhermeleobas/git/pytorch/torch/__init__.py", line 1753, in __call__
    return self.compiler_fn(model_, inputs_, **self.kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_functorch/aot_autograd.py", line 889, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/home/guilhermeleobas/git/pytorch/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/guilhermeleobas/git/pytorch/torch/_functorch/aot_autograd.py", line 507, in create_aot_dispatcher_function
    fw_metadata = run_functionalized_fw_and_collect_metadata(
  File "/home/guilhermeleobas/git/pytorch/torch/_functorch/_aot_autograd/collect_metadata_analysis.py", line 217, in inner
    curr_storage = StorageWeakRef(o.untyped_storage())
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
NotImplementedError: Cannot access storage of BatchedTensorImpl
```

guilhermeleobas replied:
Anyway, I added a test for this case but decorated it with `@unittest.expectedFailure`

zou3519 replied:
Yeah that's fine

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -3461,28 +3635,6 @@ def wrapper_fn(x):
             )
             self.assertEqual(actual, expected)
 
-    @config.patch(capture_func_transforms=True)
-    def test_vmap_illegal_op_graph_break(self):
-        counters.clear()
-
-        def bad_fn(x):
-            x.stride()
-            return x
-
-        def wrapper_fn(x):
-            return torch.func.vmap(bad_fn)(x)
-
-        x = torch.randn(3, 3, 3)
-        actual = wrapper_fn(x)
-        expected = torch.compile(wrapper_fn, backend="aot_eager", fullgraph=False)(x)
-        self.assertEqual(len(counters["graph_break"]), 1)
-        assert_dict_matches_regex(
-            self,
-            dict(counters["graph_break"]),
-            {".*Illegal getattr invocation stride in strict mode": 2},
-        )
-        self.assertEqual(actual, expected)
-
```
Can we keep this test, but instead try to return the stride in a situation where vmap changes the strides?

vmap over F.convolution sometimes changes the strides

guilhermeleobas replied:
Should the following fail on dynamo?

```python
def bad_fn(x):
    return x.view((4, 3))

def wrapper_fn(x):
    return torch.func.vmap(bad_fn)(x)

x = torch.randn(2, 3, 4)
actual = wrapper_fn(x)
expected = torch.compile(wrapper_fn, backend="aot_eager", fullgraph=False)(x)
```

zou3519 replied:
If it passes in eager mode then it should pass with Dynamo

guilhermeleobas replied:
Just saw the comment you made a while ago on @kshitij12345 PR
https://github.com/pytorch/pytorch/pull/101707#discussion_r1278030605

> We need to make sure .stride() and .storage_offset() (are there more?) are never accessed in this speculate_subgraph. Otherwise, the captured graph might be wrong, because vmap has different behavior w.r.t. strides and storage offset than regular eager-mode pytorch.

With the new vmap implementation, calling `.stride()` or `.storage_offset()` can still produce an invalid graph? I couldn't find a test case that replicates this.

zou3519 commented on test/functorch/test_eager_transforms.py:
```diff
@@ -3680,7 +3680,7 @@ def forward(self, x):
         models = [Foo() for _ in range(num_models)]
         _, params, buffers = combine_state_for_ensemble(models)
-        result = vmap(fmodel)(params, buffers, xs)
+        result = vmap(fmodel, randomness='same')(params, buffers, xs)
```
Why did this change?

guilhermeleobas replied:
Forgot to undo this change. It is not needed.

zou3519 commented on torch/_dynamo/guards.py:
```diff
@@ -359,6 +359,15 @@ def HASATTR(self, guard: Guard):
 
         self._produce_guard_code(guard, [code], provided_guarded_object=self.get(base))
 
+    def VMAP_INCREMENT_NESTING_MATCH(self, guard: Guard):
+        # Invalidate the graph if a call to vmap has been made prior to this
+        # This is super conservative as the interpreter stack may not contain
+        # vmap
+        code = [
+            "torch._C._functorch.peek_interpreter_stack() is None",
+        ]
+        self._produce_guard_code(guard, code)
```
Should the guard be a check on the current level index instead of if the interpreter stack is None?

If so, then we should change torch._C._functorch.current_level to return Optional[int] and change the guard to be a FUNCTORCH_CURRENT_LEVEL_MATCH

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2050,6 +2050,23 @@ def __init__(
             speculation_log=speculation_log,
         )
 
+        compiler_is_eager = False
+        if hasattr(compiler_fn, "compiler_name"):
+            compiler_is_eager = compiler_fn.compiler_name == "eager"
+        elif hasattr(compiler_fn, "__name__"):
+            compiler_name = compiler_fn.__name__ == "eager"
+        else:
+            # aot_eager won't match here as it has a "compiler_name" attribute
+            compiler_is_eager = "eager" in str(compiler_fn)
+
+        ci = torch._C._functorch.peek_interpreter_stack()
+        if (
+            ci is not None
+            and ci.key() == torch._C._functorch.TransformType.Vmap
+            and not compiler_is_eager
+        ):
+            unimplemented("functorch transform partial graph")
```
Let's toss this into a helper function somewhere

zou3519 replied:
Probably needs a more descriptive message :P

zou3519 commented on torch/_dynamo/trace_rules.py:
```diff
@@ -132,6 +134,32 @@
     "torch.resize_as_": SkipFilesVariable,
     "torch.resize_as_sparse_": SkipFilesVariable,
     "torch.get_default_device": TorchInGraphFunctionVariable,
+    # functorch
+    "torch._functorch.vmap._check_int_or_none": UserFunctionVariable,
```
In a follow-up PR we should ask @yanboliang if the skipfiles refactor has landed and if we can clean these up

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1019,165 +1015,17 @@ def to_python_ints(argnums):
 
 
-class FunctorchVmapHigherOrderVariable(TorchHigherOrderOperatorVariable):
+class FunctorchVmapHigherOrderVariable(UserFunctionVariable):
```
do we still need this?

zou3519 commented on torch/testing/_internal/dynamo_test_failures.py:
```diff
@@ -2467,6 +2467,50 @@
     "TestControlFlowTraced.test_map_functionalized",  # functorch/test_control_flow
     "TestControlFlowTraced.test_nested_map_cond_symbolic",  # functorch/test_control_flow
     "TestControlFlowTraced.test_nested_map_cond_real",  # functorch/test_control_flow
+    "TestJacCPU.test_against_reference_correctness_different_devices_jacfwd_cpu",  # functorch/test_eager_transforms.py
```
This is a lot of new failures. Do we know what they are?

guilhermeleobas replied:
From an oob discussion, failures are not related to this PR.

zou3519 reviewed (commented):
Only blocker is that we seem to be introducing new test failures (in dynamo_test_failures.py). Do we know what those are?

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 reviewed (approved):
LGTM. As discussed offline, we should make sure we raise a failure in tearDown (if vmap_increment_nesting was not undone) and then revisit these later -- they're bugs in dynamo, not in our vmap implementation.

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2436,7 +2437,8 @@ def tearDown(self):
                 "'torch._C._functorch._vmap_decrement_nesting()'"
             )
-            self.fail(msg)
+            warnings.warn(msg)
+            torch._C._functorch._vmap_decrement_nesting()
```
Btw, you want to make sure you undo all of the functorch stack. It's possibl that we're in multiple nested vmaps when dynamo decides to break on us

guilhermeleobas committed:
Update on "Enhance torch.vmap support from inside torch.compile"


This work rewrites vmap support in torch.compile by inlining most of
the frames into the existing FX graph. It also unlocks to PyTorch to
support features that were previously missing, such as keyword args.

Fixes: https://github.com/pytorch/pytorch/issues/114306

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
@pytorchbot merge
